### PR TITLE
SI-8144 permalinks in scaladoc

### PR DESCRIPTION
**Pull request by mkubala**
Added 'Permalink' button (visible on mouse hover) for type's members.
https://issues.scala-lang.org/browse/SI-8144

Tested under:
- Firefox 27.0.1,
- Chrome 33.0.1750.146
- Safari 6.1.2

### CODE REVIEW

**Comment by mkubala** [comment]
PLS REBUILD ALL

**Comment by VladUreche** [comment]
@mkubala please rebase on top of master:

```
git checkout master
git branch -D SI-8144
git fetch trunk
git checkout trunk/master -b SI-8144
git cherry-pick 51b9a45
git push -f SI-8144
```

This way you will only have the original commit in the pull request.

File: test/scaladoc/scalacheck/HtmlFactoryTest.scala [hunk 1]
```diff
@@ -415,36 +415,36 @@ object Test extends Properties("HtmlFactory") {
 
   property("SI-5054: Use cases should override their original members") =
      checkText("SI_5054_q1.scala")(
-       (None,"""def test(): Int""", true)
+       (None,s"""def test(): Int $permalinkText""", true)
```
Comment (VladUreche) [inline test/scaladoc/scalacheck/HtmlFactoryTest.scala]: I don't think we should test the inclusion of the peramlink text in the signatures. WDYT?

File: test/scaladoc/scalacheck/HtmlFactoryTest.scala [hunk 2]
```diff
@@ -415,36 +415,36 @@ object Test extends Properties("HtmlFactory") {
...
-      (None, """abstract def test1(): Int [use case] This takes the implicit value in scope.""", true)
+      (None, s"""abstract def test2(explicit: Int): Int $permalinkText [use case] This takes the explicit value passed.""", true),
+      (None, s"""abstract def test1(): Int $permalinkText [use case] This takes the implicit value in scope.""", true)
     )
 
   property("SI-5287: Display correct \"Definition classes\"") =
     checkText("SI_5287.scala")(
       (None,
-          """def method(): Int
+          s"""def method(): Int $permalinkText
```
Comment (VladUreche) [inline test/scaladoc/scalacheck/HtmlFactoryTest.scala]: Unfortunately it's not possible to avoid it in the next test cases ...

**Comment by VladUreche** [comment]
Thanks for the pull request Marcin! Did you sign the [Scala CLA](http://www.typesafe.com/contribute/cla/scala)?

Three problems that you should address:
- you'll need to rebase the latest commit on master (may also work with `git rebase trunk/master`) -- see previous comment
- would it be possible to replace the "permalink" text by a little image so tests don't need to be updated?
- can you please include a screenshot of the result?

For the js and css part, review by @cvogt or @heathermiller.

**Comment by VladUreche** [comment]
Regarding merging: Since 2.11 is released from the `master` branch but the PR is too late to make it into 2.11, I think it would be best to re-issue the PR in about a month from now. Ping @adriaanm.

**Comment by cvogt** [comment]
Awesome :)!  /cc: @szeiger

js & css LGTM, but I didn't try the generated docs. Somebody else please make sure the feature is usable.

**Comment by mkubala** [comment]
Thanks for review :)

I signed both Typesafe and Scala CLA.

I'll attach a screenshot (before and after replacing 'Permalink' text by an image) and maybe publish a sample scaladoc with permanent links.

**Comment by adriaanm** [comment]
If you like, you can leave the PR open -- we'll merge once the final ships. We'll likely rename master to 2.11.x and create a 2.12.x branch for the next major version, but we can do that after we merge this PR to minimize the impact on your flow.

**Comment by mkubala** [comment]
I replaced static 'Permalink' text with an image, thereby avoiding impact on test cases.
Image used as a permalink icon is my own creation and (according to the Scala CLA terms) I grant a copyright license to EPFL.

Also code responsible for permalink html element creation was moved from Page into more specialized HtmlPage.

You can preview results at http://scaladoc-permalinks.herokuapp.com/index.html#package

@adriaanm I'll leave this PR open. I'll be thankful if you merge this PR before branch rename.

**Comment by VladUreche** [comment]
Marcin, thanks for the updated PR! LGTM for the scaladoc changes, but you'll need to fix the test:

```
/localhome/jenkins/c/workspace/pr-scala-test/scala/build.xml:1488: The following error occurred while executing this line:
/localhome/jenkins/c/workspace/pr-scala-test/scala/build-ant-macros.xml:723: Test suite finished with 1 case failing:
fail - scalacheck/HtmlFactoryTest.scala  [compilation failed]% scalac HtmlFactoryTest.scala
HtmlFactoryTest.scala:669: error: value o is not a member of scala.xml.NodeSeq
    val noAuthors = createTemplates("SI-4014_0.scala")("Foo.html")                                             o
                                                                                                               ^
one error found
```

File: test/scaladoc/scalacheck/HtmlFactoryTest.scala [hunk 3]
```diff
@@ -667,7 +666,7 @@ object Test extends Properties("HtmlFactory") {
   }
 
   property("SI-4014: Scaladoc omits @author: no authors") = {
-    val noAuthors = createTemplates("SI-4014_0.scala")("Foo.html")
+    val noAuthors = createTemplates("SI-4014_0.scala")("Foo.html")                                             o
```
Comment (VladUreche) [inline test/scaladoc/scalacheck/HtmlFactoryTest.scala]: @mkubala Here's the problem: there's a stray "o" at the end of the line.

Comment (mkubala) [inline test/scaladoc/scalacheck/HtmlFactoryTest.scala]: That is a Murphy's law in real life. The only one time when I was too confident and skip reading the output from git diff before commit.. It's so embarrassing :S

**Comment by som-snytt** [comment]
That's pretty funny.  People are always complaining about trailing whitespace, but never about trailing o's.

**Comment by mkubala** [comment]
[hipsterMode]trailing whitespaces are so mainstream..[/hipsterMode] ;)
But seriously - I'm waiting for ant to finish its job and I'll commit appropriate fix.

**Comment by mkubala** [comment]
PLS REBUILD ALL

**Comment by mkubala** [comment]
I append test fix as a separate commit, but I can squash them before merge if needed.

**Comment by VladUreche** [comment]
@mkubala Please do squash everything into a commit, since any commit in the history should compile well. Thanks and sorry for so much going back and forth.

**Comment by mkubala** [comment]
Commits already squashed 
(I wrote this comment because I'm not sure if adding commits to the PR results in participants notification..)

**Comment by VladUreche** [comment]
LGTM! Thanks a lot for your work Marcin! And sorry for requiring so much engineering on your side to have the PR ready, but that's the only way to keep the compiler in a good shape.

Now we only need to wait until 2.11 is released and then Adriaan will merge the patch.

**Comment by retronym** [comment]
Master branch is still 2.11.1-SNAPSHOT. We will branch off 2.12.0-SNAPSHOT shortly. So this PR has landed in the right place.

**Pull request merged**